[SPARK-49113] Dont assert on translation bugs - silently swallow the exception

milastdbx opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'common/utils/src/main/resources/error/README.md'.
-->


### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
In this PR I propose that we do not assert and fail queries when `V2ExpressionBuilder` returns class that is not predicate. Instead I suggest that we silently swallow the exception, add log, and continue execution of query.

`V2ExpressionBuilder` has no compiler gurantees that `Predicate` will be returned. With this in mind, its very easy to add regressions that will suddenly start failing queries, without a workaround for users, other than code change.

In this PR, I suggest we add a flag, which when enabled, will absorb this bugs in `V2ExpressionBuilder`.

It is as if we are saying:
 - If we are trying to translate catalyst expression to `V2Expression` and are requesting `Predicate`, if predicate is not returned act as if we didn't know how to translate expression at all.

We return None, and rest of the system handles it as if we didn't know to translate to `V2Expression`

More details in JIRA ticket

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

Queries that failed because of pushdown failure, will now skip pushing down offending predicate, rather than failing the whole query.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->

Unit tests.

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No

milastdbx committed:
remove assert from datasource v2 strategy

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1487,6 +1487,13 @@ object SQLConf {
+      .internal()
+      .doc("Enable skipping assert when expression in not translated to predicate.")
+      .booleanConf
+      .createWithDefault(true)
```
In test mode, we should not skip the assertions so that we can still capture the bug.

gatorsmile replied:
.createWithDefault(Utils.isTesting)

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala:
```diff
@@ -650,10 +651,22 @@ private[sql] object DataSourceV2Strategy extends Logging {
+        !exprOpt.get.isInstanceOf[Predicate]) {
+        logWarning(log"Predicate expected but got class: ${MDC(EXPR, exprOpt.get.describe())}")
+        None
+      } else { exprOpt }
```
the indentation is not right.  See the guideline 

https://github.com/databricks/scala-style-guide#spacing-and-indentation

milastdbx replied:
I fail see whats wrong here ?
Is its two space identation for after codeblock within if - right ?
and 2 spaces for each new line, and 2 spaces for each if condition ?

Are you referring to else branch should be in new line ?

gatorsmile replied:
In our code base, the suggested indentation should be like this 

```
} else { 
  exprOpt 
}
```

milastdbx replied:
Got it!

Doc you shared is not specific around this tho. Even has different example:
<img width="1145" alt="image" src="https://github.com/user-attachments/assets/43c9a0c0-0733-4ecd-b25b-68abadea06b3">

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/connector/PushablePredicateSuite.scala:
```diff
@@ -0,0 +1,49 @@
+      }
+    }
+  }
+}
```
add one more line

gatorsmile reviewed (changes requested):
> Does this PR introduce any user-facing change?
Queries that used to fail because of a bug in translation from catalyst to V2 expressions will now work by skipping expression which is not predicate. And also add a log.

This is still unclear. Basically, your changes are trying to simply skip the potential pushdown, instead of failing the queries. 

However, in the test mode, we should still allow the query to fail. This will help us identify and capture any underlying bugs, rather than silently bypassing them. Ignoring these bugs could lead to missed opportunities for predicate pushdown, which might negatively impact query performance.

WDYT?

milastdbx committed:
moved logic to v2 expression builder

milastdbx committed:
refactored

milastdbx commented:
> This is still unclear. Basically, your changes are trying to simply skip the potential pushdown, instead of failing the queries.
However, in the test mode, we should still allow the query to fail. This will help us identify and capture any underlying bugs, rather than silently bypassing them. Ignoring these bugs could lead to missed opportunities for predicate pushdown, which might negatively impact query performance.
WDYT?

Agreed. Updated description and added test assert. Also refactored code to have cleaner API as per agreement with @sigmod

milastdbx committed:
test fix

milastdbx committed:
refactored code a bit

milastdbx committed:
scalastyle

milastdbx committed:
Merge branch 'master' into dev/milast/removeAssert

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala:
```diff
@@ -25,15 +27,44 @@ import org.apache.spark.sql.connector.expressions.{Cast => V2Cast, Expression =>
+      }
+
+      modifiedExprOpt.map { v =>
+        assert(v.isInstanceOf[V2Predicate])
```
Enhance the assert statement by incorporating additional context, such as variable values or descriptive messages, to facilitate easier debugging?

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/connector/PushablePredicateSuite.scala:
```diff
@@ -0,0 +1,56 @@
+  test("PushablePredicate throws") {
+    withSQLConf(SQLConf.DATA_SOURCE_DONT_ASSERT_ON_PREDICATE.key -> "false") {
+      intercept[java.lang.AssertionError] {
+        val pushable = PushablePredicate.unapply(Literal.create("string"))
```
remove val?

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala:
```diff
@@ -25,15 +27,44 @@ import org.apache.spark.sql.connector.expressions.{Cast => V2Cast, Expression =>
+
+        // When isPredicate is specified we always expect to receive predicate.
+        // If this is not the case, and we received something that is not a predicate -
+        // don't fail query, act as if we didn't know how to translate it at all.
```
How about ? ```
    // If a predicate is expected but the translation yields something else,
    // log a warning and proceed as if the translation was not possible.```

gatorsmile commented:
@gengliangwang  can you take a look?

milastdbx committed:
addressing comments

gengliangwang commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/util/V2ExpressionBuilder.scala:
```diff
@@ -25,15 +27,43 @@ import org.apache.spark.sql.connector.expressions.{Cast => V2Cast, Expression =>
 
   def build(): Option[V2Expression] = generateExpression(e, isPredicate)
 
+  def buildPredicate(): Option[V2Predicate] = {
```
let's simplify the logic a little bit:
```
  def buildPredicate(): Option[V2Predicate] = {
    if (isPredicate) {
      val translated = build()

      if (translated.isDefined) {
        if (SQLConf.get.getConf(SQLConf.DATA_SOURCE_DONT_ASSERT_ON_PREDICATE)) {
          logWarning(log"Predicate expected but got class: ${MDC(EXPR, translated.get.describe())}")
          assert(translated.get.isInstanceOf[V2Predicate])
          Some(translated.asInstanceOf[V2Predicate])
        } else {
          translated.collectFirst({ case p: V2Predicate => p })
        }
      } else {
        None
      }
    } else {
      None
    }
  }
```

milastdbx replied:
these two codes don't do the same thing.

What in current code do you find complex ? IMO current code might be a bit less in scala spirit but is fairly linear

gengliangwang commented on sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:
```diff
@@ -1487,6 +1487,13 @@ object SQLConf {
+      .internal()
+      .doc("Enable skipping assert when expression in not translated to predicate.")
+      .booleanConf
+      .createWithDefault(!Utils.isTesting)
```
shall we create the default value as false, and then always enable the check if `Utils.isTesting` is true?

milastdbx replied:
flag is - DONT_ASSERT and we want to assert in testing hence default `!Utils.isTesting`.

gengliangwang commented:
Thanks, merging to master
